lib/fastlane/plugin/wpmreleasetoolkit/helper/ios/ios_strings_file_validation_helper.rb

```diff
@@ -0,0 +1,124 @@
+        # which they are defined.
+        #
+        # @param [String] file The path to the file to inspect.
+        # @return [<Hash>] A `Hash` with keys the key values read from the
+        #         `.strings` file. The value of each entry is an array with the
+        #         line numbers where the key occurs.
+        def self.get_keys_from_strings(file:)
+          keys_with_lines = Hash.new([])
+
+          state = State.new(context: :root, buffer: StringIO.new, in_escaped_ctx: false, found_key: nil)
+
+          File.readlines(file).each_with_index do |line, line_no|
+            line.chars.each_with_index do |c, col_no|
+              # Handle escaped characters at a global level. This is more
+              # straightforward than having a `TRANSITIONS` table that account
+              # for it.
+              if state.in_escaped_ctx || c == '\\'
+                # Just because we check for escaped characters at the global
+                # level, it doesn't mean we allow them in every context.
+                allowed_contexts_for_escaped_characters = %i[in_quoted_key in_quoted_value in_block_comment in_line_comment]
+                raise "Found escaped character outside of allowed contexts on line #{line_no + 1} (current context: #{state.context})" unless allowed_contexts_for_escaped_characters.include?(state.context)
+
+                state.buffer.write(c) if state.context == :in_quoted_key
+                state.in_escaped_ctx = !state.in_escaped_ctx
+                next
+              end
+
+              # Look at the transitions table for the current context, and find
+              # the first transition matching the current character
+              (_, next_context) = TRANSITIONS[state.context].find { |regex, _| c.match?(regex) } || [nil, nil]
+              raise "Invalid character `#{c}` found on line #{line_no + 1}, col #{col_no + 1}" if next_context.nil?
+
+              state.context = next_context.is_a?(Proc) ? next_context.call(state, c) : next_context
+              next unless state.found_key
+
+              # If we just exited the :in_quoted_key context and thus have found
+              # a new key, process it
+              key = state.found_key.dup
+              state.found_key = nil
+
+              keys_with_lines[key] += [line_no + 1]
+            end
+          end
+
+          keys_with_lines
+        end
+
+        # Inspects the given `.strings` file for duplicated keys, returning them
+        # if any.
+        #
+        # @param [String] file The path to the file to inspect.
+        # @return [Array<Hash>] Array of all the duplicated keys. Each entry is a
+        #         a `Hash` with keys: `key` for the key value and `lines` with an
+        #         array of line numbers at which the key occurs.
+        def self.find_duplicated_keys(file:)
+          get_keys_from_strings(file: file).keep_if { |_, lines| lines.count > 1 }
+        end
```

Comment:
Given how simple this one has become, I wonder if it's still necessary to have 2 separate methods now.

Could make for some nice tidying to remove the `get_keys_from_strings(file:)` method, and move its whole implementation directly into this `find_duplicated_keys` instead of this one-liner of one calling the other, in order to merge the 2 methods into one, wdyt?